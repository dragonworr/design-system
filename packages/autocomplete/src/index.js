import React from 'react'
import PropTypes from 'prop-types'
import Downshift from 'downshift'
import styled from 'styled-components'
import {
  Flex,
  Label as PclnLabel,
  Input as PclnInput
} from 'pcln-design-system'

export const AutocompleteContext = React.createContext()

export class Autocomplete extends React.Component {
  static Label = Label
  static Input = Input
  static Menu = Menu
  static Item = Item

  static propTypes = {
    match: PropTypes.func
  }

  static defaultProps = {
    match: () => true
  }

  render() {
    const { children, ...props } = this.props

    return (
      <Downshift
        {...props}
        children={state => (
          <div>
            <AutocompleteContext.Provider
              value={{ ...props, ...state }}
              children={children}
            />
          </div>
        )}
      />
    )
  }
}

export const withAutocomplete = (Component, mapProps) =>
  React.forwardRef((props, ref) => (
    <AutocompleteContext.Consumer
      children={state => (
        <Component ref={ref} {...mapProps({ ...props, ...state })} {...props} />
      )}
    />
  ))

export const Label = withAutocomplete(PclnLabel, ({ getLabelProps }) =>
  getLabelProps()
)

export const Input = withAutocomplete(PclnInput, ({ getInputProps }) =>
  getInputProps()
)

export const Menu = ({ children, ...props }) => (
  <AutocompleteContext.Consumer
    children={({ match, isOpen, getMenuProps, inputValue }) =>
      isOpen ? (
        <div
          {...getMenuProps()}
          {...props}
          children={React.Children.toArray(children)
            .filter(el => match(el.props.item, inputValue))
            .map((el, index) => React.cloneElement(el, { index }))}
        />
      ) : (
        false
      )
    }
  />
)

const ItemRoot = styled(Flex)({
  '&[data-selected]': {
    backgroundColor: 'blue'
  },
  '&[data-highlighted]': {
    backgroundColor: 'yellow'
  }
})

export const Item = withAutocomplete(
  ItemRoot,
  ({
    item,
    index,
    getItemProps,
    selectedItem,
    highlightedIndex,
    children,
    ...props
  }) => ({
    px: 2,
    py: 1,
    ...getItemProps({ item, index, ...props }),
    children,
    'data-selected': selectedItem === item ? true : undefined,
    'data-highlighted': highlightedIndex === index ? true : undefined
  })
)
