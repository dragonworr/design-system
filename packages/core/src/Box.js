import styled from 'styled-components'
import { space, width, textAlign } from 'styled-system'
import { applyVariations, color, deprecatedColorValue } from './utils'

const Box = styled.div`
  ${space} ${width} ${textAlign}
  ${applyVariations('Box')}
  ${color}
`

Box.displayName = 'Box'

Box.propTypes = {
  ...space.propTypes,
  ...width.propTypes,
  color: deprecatedColorValue(),
  bg: deprecatedColorValue(),
  ...textAlign.propTypes
}

export default Box
