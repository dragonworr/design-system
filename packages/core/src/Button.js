import styled, { css } from 'styled-components'
import PropTypes from 'prop-types'
import { width, space } from 'styled-system'
import {
  mapProps,
  deprecatedPropType,
  applyVariant,
  getPaletteColor,
  getTextColorOn
} from './utils'

const size = props => {
  switch (props.size) {
    case 'small':
      return {
        fontSize: `${props.theme.fontSizes[0]}px`,
        padding: '7px 12px'
      }
    case 'medium':
      return {
        fontSize: `${props.theme.fontSizes[1]}px`,
        padding: '9.5px 18px'
      }
    case 'large':
      return {
        fontSize: `${props.theme.fontSizes[2]}px`,
        padding: '12px 22px'
      }
    default:
      return {
        fontSize: `${props.theme.fontSizes[1]}px`,
        padding: '9.5px 18px'
      }
  }
}

const variants = {
  fill: css`
    background-color: ${getPaletteColor('base')};
    color: ${getTextColorOn('base')};

    &:hover {
      background-color: ${getPaletteColor('dark')};
      color: ${getTextColorOn('dark')};
    }
  `,
  outline: css`
    color: ${getPaletteColor('base')};
    box-shadow: inset 0 0 0 2px ${getPaletteColor('base')};
    background-color: transparent;

    &:hover {
      color: ${getPaletteColor('dark')};
      box-shadow: inset 0 0 0 2px ${getPaletteColor('dark')};
      background-color: transparent;
    }
  `,
  disabled: css`
    background-color: ${getPaletteColor('light')};
    color: ${getTextColorOn('light')};
    cursor: default;

    &:hover {
      background-color: null;
    }
  `
}

const Button = mapProps(({ fullWidth, ...props }) => ({
  width: fullWidth ? 1 : undefined,
  ...props
}))(styled.button`
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  text-decoration: none;
  font-family: inherit;
  font-weight: 600;
  line-height: 1.5;
  cursor: pointer;
  border-radius: ${props => props.theme.radius};
  border-width: 0;
  border-style: solid;

  &:disabled {
    opacity: 0.25;
  }

  ${applyVariant('Button', variants)}

  ${width} ${size} ${space};
`)

Button.propTypes = {
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  ...width.propTypes,
  ...space.propTypes,
  fullWidth: deprecatedPropType('width'),
  variant: PropTypes.oneOf(Object.keys(variants)),
  color: PropTypes.string
}

Button.defaultProps = {
  color: 'primary',
  variant: 'fill'
}

Button.displayName = 'Button'

export default Button
