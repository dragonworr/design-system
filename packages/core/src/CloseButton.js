import React from 'react'
import PropTypes from 'prop-types'
import IconButton from './IconButton'
import { Close } from 'pcln-icons'
import { refPropType } from './utils'

const CloseButton = props => (
  <IconButton {...props} icon={<Close size={props.size} />} />
)

CloseButton.defaultProps = {
  size: 24,
  title: 'close'
}

CloseButton.propTypes = {
  onClick: PropTypes.func,
  size: PropTypes.number,
  dsRef: refPropType,
  title: PropTypes.string
}

CloseButton.displayName = 'CloseButton'

export default CloseButton
