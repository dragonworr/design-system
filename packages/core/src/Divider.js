import styled from 'styled-components'
import { applyVariations, getPaletteColor, deprecatedColorValue } from './utils'
import { space, width, borderColor } from 'styled-system'

const Divider = styled.hr`
  border: 0;
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-color: ${({ color, borderColor, ...props }) =>
      getPaletteColor('base')({ color: borderColor || color, ...props })}
    ${space} ${width};
  ${applyVariations('Divider')}
`

Divider.displayName = 'Divider'

Divider.defaultProps = {
  color: 'border',
  ml: 0,
  mr: 0
}

Divider.propTypes = {
  ...space.propTypes,
  ...width.propTypes,
  ...borderColor.propTypes,
  color: deprecatedColorValue()
}

export default Divider
