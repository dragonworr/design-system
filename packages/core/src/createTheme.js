export default theme => {
  return {
    ...theme,
    componentStyles: {
      buttonPrimary: {
        backgroundColor: theme.palette.primary,
        color: theme.colors.white,
        '&:hover': {
          backgroundColor: theme.colors.darkBlue
        },
        ...((theme.componentStyles || {}).buttonPrimary || {})
      },
      buttonSecondary: {
        backgroundColor: theme.palette.secondary,
        color: theme.colors.white,
        '&:hover': {
          backgroundColor: theme.colors.darkGreen
        },
        ...((theme.componentStyles || {}).buttonSecondary || {})
      },
      buttonDefault: {
        backgroundColor: theme.palette.primary,
        color: theme.colors.white,
        '&:hover': {
          backgroundColor: theme.colors.darkBlue
        },
        ...((theme.componentStyles || {}).buttonDefault || {})
      },
      buttonDisabled: {
        backgroundColor: theme.colors.lightGray,
        color: theme.palette.textPrimary,
        cursor: 'default',
        '&:hover': {
          backgroundColor: null
        },
        ...((theme.componentStyles || {}).buttonDisabled || {})
      },
      buttonNegative: {
        backgroundColor: theme.palette.error,
        color: theme.colors.white,
        '&:hover': {
          backgroundColor: theme.colors.darkRed
        },
        ...((theme.componentStyles || {}).buttonNegative || {})
      },
      linkPrimary: {
        color: theme.palette.primary,
        '&:hover': {
          textDecoration: 'underline'
        },
        ...((theme.componentStyles || {}).linkPrimary || {})
      },
      linkSecondary: {
        color: theme.palette.secondary,
        '&:hover': {
          textDecoration: 'underline'
        },
        ...((theme.componentStyles || {}).linkSecondary || {})
      },
      linkTitle: {
        color: theme.colors.red,
        '&:hover': {
          textDecoration: 'underline'
        },
        ...((theme.componentStyles || {}).linkTitle || {})
      },
      linkNavigation: {
        color: theme.palette.textPrimary,
        '&:hover': {
          textDecoration: 'underline'
        },
        ...((theme.componentStyles || {}).linkNavigation || {})
      }
    }
  }
}
