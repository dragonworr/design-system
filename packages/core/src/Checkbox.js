import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import theme from './theme'
import Icon from './Icon'
import Box from './Box'
import { getPaletteColor } from './utils'

const Checkbox = props => {
  const { disabled, size } = props
  return (
    <CheckBoxWrapper color={props.color} disabled={disabled}>
      <StyledInput type="checkbox" {...props} />
      <Icon name="BoxChecked" size={size} data-name="checked" />
      <Icon name="BoxEmpty" size={size} data-name="empty" />
    </CheckBoxWrapper>
  )
}

const CheckBoxWrapper = styled(Box)`
  display: inline-block;
  position: relative;
  vertical-align: middle;
  cursor: pointer;
  color: ${props =>
    props.disabled
      ? getPaletteColor('border.base')(props)
      : getPaletteColor('border.light')(props)};

  svg[data-name='checked'] {
    display: none;
  }

  > input:checked {
    & ~ svg[data-name='checked'] {
      display: inline-block;
      color: ${props =>
        props.disabled
          ? getPaletteColor('border.base')(props)
          : getPaletteColor('base')(props)};
    }

    & ~ svg[data-name='empty'] {
      display: none;
    }
  }
`

CheckBoxWrapper.defaultProps = {
  theme
}

const StyledInput = styled.input`
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
`

Checkbox.displayName = 'Checkbox'

Checkbox.propTypes = {
  id: PropTypes.string.isRequired,
  size: PropTypes.number,
  onChange: PropTypes.func.isRequired,
  color: PropTypes.string
}

Checkbox.defaultProps = {
  size: 20,
  color: 'primary'
}

export default Checkbox
