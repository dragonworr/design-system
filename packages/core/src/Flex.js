import styled from 'styled-components'
import Box from './Box'
import PropTypes from 'prop-types'
import {
  space,
  width,
  color,
  alignItems,
  justifyContent,
  flexWrap,
  flexDirection
} from 'styled-system'
import theme from './theme'
import { mapProps, deprecatedPropType } from './utils'

const Flex = mapProps(({ wrap, align, justify, ...props }) => ({
  flexWrap: wrap ? 'wrap' : undefined,
  alignItems: align,
  justifyContent: justify,
  ...props
}))(styled(Box)`
  display: flex;
  ${alignItems} ${justifyContent}
  ${flexDirection}
  ${flexWrap}
`)

Flex.propTypes = {
  ...space.propTypes,
  ...width.propTypes,
  // color: PropTypes.string,
  ...alignItems.propTypes,
  ...justifyContent.propTypes,
  ...flexWrap.propTypes,
  ...flexDirection.propTypes,
  // deprecated
  wrap: deprecatedPropType('flexWrap'),
  align: deprecatedPropType('alignItems'),
  justify: deprecatedPropType('justifyContent')
}

Flex.displayName = 'Flex'

export default Flex
