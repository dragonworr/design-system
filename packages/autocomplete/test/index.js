import React from 'react'
import TestRenderer from 'react-test-renderer'
import us from 'us'
import { Autocomplete } from '../src'

describe('Autocomplete', () => {
  test('renders', () => {
    const json = TestRenderer.create(
      <Autocomplete>
        <Autocomplete.Label>Test</Autocomplete.Label>
        <Autocomplete.Input />
        <Autocomplete.Menu>
          {us.STATES.map(state => (
            <Autocomplete.Item
              key={state.name}
              item={state}
              children={state.name}
            />
          ))}
        </Autocomplete.Menu>
      </Autocomplete>
    ).toJSON()
    expect(json).toMatchSnapshot()
  })
})
