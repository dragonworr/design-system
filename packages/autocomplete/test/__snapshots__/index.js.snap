// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-labelledby="downshift-0-label"
  aria-owns={null}
  role="combobox"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <label
    className="sc-cMljjf fTtdJC"
    color="gray"
    fontSize="10px"
    fontWeight="bold"
    htmlFor="downshift-0-input"
    id="downshift-0-label"
  >
    Test
  </label>
  <input
    aria-activedescendant={null}
    aria-autocomplete="list"
    aria-controls={null}
    aria-labelledby="downshift-0-label"
    autoComplete="off"
    className="sc-jWBwVP iFOPYV"
    id="downshift-0-input"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    value=""
  />
  <div
    aria-labelledby="downshift-0-label"
    className="sc-kjoXOD cTsYwp sc-kGXeez dpMHNO sc-bwzfXH cGrtGY"
    height="256px"
    id="downshift-0-menu"
    role="listbox"
  />
</div>
`;
